include/swift/SIL/GraphFunctionDeviceInfo.h

```diff
@@ -29,36 +31,61 @@ namespace tf {
 class GraphOperationBuilder;
 struct GraphOperationInfo;
 
-/// The device of a tfop instruction (and its output tensors, if any).
+/// The device type of a tfop instruction (and its output tensors, if any).
 enum class DeviceType {
   INVALID,
   CPU,
   GPU,
   TPU,
   /// Indicates this instruction should run on all devices in
-  /// `GraphFunctionDeviceInfo::usedDeviceTypes`. For example, a promoted
+  /// `GraphFunctionDeviceInfo::usedDeviceIds`. For example, a promoted
   /// scalar will run on all such devices, in case it is a loop iteration count
   /// and the loop runs on all devices.
   ALL,
 };
 
-static const char TF_DEFAULT_CPU_DEVICE[] =
-    "/job:localhost/replica:0/task:0/device:CPU:0";
-static const char TF_DEFAULT_GPU_DEVICE[] =
-    "/job:localhost/replica:0/task:0/device:GPU:0";
+/// The device id represents a local device, with two components: device type
+/// and index. An example is "GPU:3".
+// TODO: consider adding a worker (task) component to support remote devices.
+struct DeviceId {
+  DeviceType type;
+  unsigned index;
+
+  /// Map between a DeviceId object and a 64-bit int. The latter is used when we
+  /// use DeviceId as the key type of a set/map container.
+  // We use 32 bits to encode `type`, which is wasteful.
+  uint64_t encode() const { return ((uint64_t)type << 32) | index; }
```

Comment:
I think it's better to use the general `llvm::DenseMapInfo` trait to be consistent with the rest of the code base. That would also allow you to use `DeviceID` as the key.

Reply:
Switched to using `llvm::DenseMapInfo`. Indeed the code is more consistent and the call-sites are simpler this way. Nice suggestion!